WalletWasabi/Crypto/StringCipher.cs

```diff
@@ -1,127 +1,141 @@
-		private const int KeySize = 128;
+		private const int AesBlockByteSize = 128 / 8;
 
-		// This constant determines the number of iterations for the password bytes generation function.
-		private const int DerivationIterations = 1000;
+		private const int PasswordSaltByteSize = 128 / 8;
+		private const int PasswordByteSize = 256 / 8;
+		private const int PasswordIterationCount = 100_000;
 
-		public static string Encrypt(string plainText, string passPhrase)
+		private const int SignatureByteSize = 256 / 8;
+
+		private const int MinimumEncryptedMessageByteSize =
+			PasswordSaltByteSize + // Auth salt
+			PasswordSaltByteSize + // Key salt
+			AesBlockByteSize + // IV
+			AesBlockByteSize + // Cipher text min length
+			SignatureByteSize; // Signature tag
+
+		private static readonly Encoding StringEncoding = Encoding.UTF8;
+		private static readonly RandomNumberGenerator Random = RandomNumberGenerator.Create();
+
+		public static string Encrypt(string toEncrypt, string password)
 		{
-			// Salt is randomly generated each time, but is prepended to encrypted cipher text
-			// so that the same Salt value can be used when decrypting.
-			byte[] salt = Generate128BitsOfRandomEntropy();
-			byte[] iv;
-			byte[] cipherTextBytes;
-			byte[] key;
-			var plainTextBytes = Encoding.UTF8.GetBytes(plainText);
-
-			using (var password = new Rfc2898DeriveBytes(passPhrase, salt, DerivationIterations))
+			byte[] keySalt = GenerateRandomBytes(PasswordSaltByteSize);
+			byte[] key = GetKey(password, keySalt);
+			byte[] iv = GenerateRandomBytes(AesBlockByteSize);
+
+			byte[] cipherText;
+			using (Aes aes = CreateAes())
+			using (ICryptoTransform encryptor = aes.CreateEncryptor(key, iv))
 			{
-				key = password.GetBytes(KeySize / 8);
-				using var aes = CreateAES();
-				aes.GenerateIV();
-				iv = aes.IV;
-				using var encryptor = aes.CreateEncryptor(key, iv);
-				using var memoryStream = new MemoryStream();
-				using (var cryptoStream = new CryptoStream(memoryStream, encryptor, CryptoStreamMode.Write))
-				{
-					cryptoStream.Write(plainTextBytes, 0, plainTextBytes.Length);
-					cryptoStream.FlushFinalBlock();
-					cryptoStream.Close();
-				}
-				cipherTextBytes = memoryStream.ToArray();
+				byte[] plainText = StringEncoding.GetBytes(toEncrypt);
+				cipherText = encryptor.TransformFinalBlock(plainText, 0, plainText.Length);
 			}
 
-			using (var memoryStream = new MemoryStream())
-			{
-				using (var writer = new BinaryWriter(memoryStream))
-				{
-					writer.Write(salt);
-					writer.Write(iv);
-					using (var hmac = new HMACSHA256(key))
-					{
-						var authenticationCode = hmac.ComputeHash(iv.Concat(cipherTextBytes).ToArray());
-						writer.Write(authenticationCode);
-					}
-					writer.Write(cipherTextBytes);
-					writer.Flush();
-				}
+			// Sign.
+			byte[] authKeySalt = GenerateRandomBytes(PasswordSaltByteSize);
+			byte[] authKey = GetKey(password, authKeySalt);
```

Comment:
I've been reading more about the best practices and this is the recommend way to do it, i mean, using a different key for mac and encryption.